// @react-ol/fiber/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "catalogue", {
    enumerable: true,
    get: function() {
        return _catalogue.catalogue;
    }
});
Object.defineProperty(exports, "extend", {
    enumerable: true,
    get: function() {
        return _catalogue.extend;
    }
});
Object.defineProperty(exports, "MapContext", {
    enumerable: true,
    get: function() {
        return _context.MapContext;
    }
});
Object.defineProperty(exports, "MapProvider", {
    enumerable: true,
    get: function() {
        return _context.MapProvider;
    }
});
Object.defineProperty(exports, "MapConsumer", {
    enumerable: true,
    get: function() {
        return _context.MapConsumer;
    }
});
Object.defineProperty(exports, "useMap", {
    enumerable: true,
    get: function() {
        return _context.useMap;
    }
});
Object.defineProperty(exports, "Map", {
    enumerable: true,
    get: function() {
        return _map.Map;
    }
});
var _exportNames = {
    catalogue: true,
    extend: true,
    MapContext: true,
    MapProvider: true,
    MapConsumer: true,
    useMap: true,
    Map: true
};
var _catalogue = require("./catalogue");
var _context = require("./context");
var _map = require("./map");
var _types = _interopRequireWildcard(require("./types"));
Object.keys(_types).forEach(function(key) {
    if (key === "default" || key === "__esModule") return;
    if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
    if (key in exports && exports[key] === _types[key]) return;
    Object.defineProperty(exports, key, {
        enumerable: true,
        get: function() {
            return _types[key];
        }
    });
});
function _interopRequireWildcard(obj) {
    if (obj && obj.__esModule) {
        return obj;
    } else {
        var newObj = {};
        if (obj != null) {
            for(var key in obj){
                if (Object.prototype.hasOwnProperty.call(obj, key)) {
                    var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {};
                    if (desc.get || desc.set) {
                        Object.defineProperty(newObj, key, desc);
                    } else {
                        newObj[key] = obj[key];
                    }
                }
            }
        }
        newObj.default = obj;
        return newObj;
    }
}
